File: src/backend/auth/TableSessions.cpp
#include "backend/auth/TableSessions.h"
#include <ctime>
#include <iostream>
#include "database/DataBase.h"

namespace oink_judge::backend::auth {

using DataBase = database::DataBase;

TableSessions &TableSessions::instance() {
    static TableSessions instance;
    return instance;
}

bool TableSessions::add_session(const Session &session) {
    try {
        DataBase::instance().execute("sessions__add_session", session.get_session_id(), session.get_username(), session.get_expire_at());
        return true;
    } catch (const pqxx::sql_error& e) {
        std::cerr << "SQL error: " << e.what() << std::endl;
        return false;
    }
}

bool TableSessions::remove_session(const Session &session) {
    try {
        DataBase::instance().execute("sessions__remove_session", session.get_session_id());
        return true;
    } catch (const pqxx::sql_error& e) {
        std::cerr << "SQL error: " << e.what() << std::endl;
        return false;
    }
}

bool TableSessions::remove_session(const std::string &session_id) {
    try {
        DataBase::instance().execute("sessions__remove_session", session_id);
        return true;
    } catch (const pqxx::sql_error& e) {
        std::cerr << "SQL error: " << e.what() << std::endl;
        return false;
    }
}


bool TableSessions::clear_all_user_sessions(const std::string &username) {
    try {
        DataBase::instance().execute("sessions__remove_by_username", username);
        return true;
    } catch (const pqxx::sql_error& e) {
        std::cerr << "SQL error: " << e.what() << std::endl;
        return false;
    }
}

std::string TableSessions::whose_session(const std::string &session_id) const {
    try {
        pqxx::result res = DataBase::instance().execute("sessions__select_session", session_id);

        if (res.empty()) return "";
        if (res.size() > 1) {
            throw std::runtime_error("Session ID must contain only one session");
        }

        auto stored_username = res[0]["username"].as<std::string>();
        auto stored_expire_at = res[0]["expire_at"].as<time_t>();

        if (_is_expired(stored_expire_at)) {
            DataBase::instance().execute("sessions__remove_session", session_id);
            return "";
        }

        return stored_username;
    } catch (const pqxx::sql_error& e) {
        std::cerr << "SQL error: " << e.what() << std::endl;
        return "";
    }
}

TableSessions::TableSessions() {
    const std::string create_sql = "CREATE TABLE IF NOT EXISTS sessions ("
                                       "id TEXT PRIMARY KEY, "
                                       "username TEXT, "
                                       "expire_at INTEGER);";

    DataBase::instance().execute_sql(create_sql);

    const std::string add_session_sql = "INSERT INTO sessions (id, username, expire_at) VALUES ($1, $2, $3)";

    const std::string remove_session_sql = "DELETE FROM sessions WHERE id = $1";

    const std::string remove_sessions_by_username_sql = "DELETE FROM sessions WHERE username = $1";

    const std::string select_session_sql = "SELECT username, expire_at FROM sessions WHERE id = $1";


    DataBase::instance().prepare_statement("sessions__add_session", add_session_sql);
    DataBase::instance().prepare_statement("sessions__remove_session", remove_session_sql);
    DataBase::instance().prepare_statement("sessions__remove_by_username", remove_sessions_by_username_sql);
    DataBase::instance().prepare_statement("sessions__select_session", select_session_sql);
}


bool TableSessions::_is_expired(time_t expire_at) {
    return time(nullptr) > expire_at;
}

} // namespace oink_judge::backend::auth

File: app/services/auth/table_sessions.py
import time

from oink_judge.pybind11_database import (
    ConnectionPool,
    execute,
    execute_sql,
)

from app.services.auth.session import Session


class TableSessions:
    _instance: "TableSessions | None" = None

    def __init__(self) -> None:
        self._initialized = False

    @classmethod
    def instance(cls) -> "TableSessions":
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    async def _require_initialized(self) -> None:
        if not self._initialized:
            await self.initialize()

    async def initialize(self) -> None:
        if self._initialized:
            return

        await execute_sql(
            "CREATE TABLE IF NOT EXISTS sessions ("
            "id TEXT PRIMARY KEY, "
            "username TEXT, "
            "expire_at INTEGER);"
        )

        pool = ConnectionPool.instance()
        pool.prepare_statement(
            "sessions__add_session",
            "INSERT INTO sessions (id, username, expire_at) VALUES ($1, $2, $3)",
        )
        pool.prepare_statement(
            "sessions__remove_session",
            "DELETE FROM sessions WHERE id = $1",
        )
        pool.prepare_statement(
            "sessions__remove_by_username",
            "DELETE FROM sessions WHERE username = $1",
        )
        pool.prepare_statement(
            "sessions__select_session",
            "SELECT username, expire_at FROM sessions WHERE id = $1",
        )
        self._initialized = True

    async def add_session(self, session: Session) -> bool:
        await self._require_initialized()
        await execute(
            "sessions__add_session",
            session.session_id,
            session.username,
            session.expire_at,
        )
        return True

    async def remove_session(self, session_id: str) -> bool:
        await self._require_initialized()
        await execute("sessions__remove_session", session_id)
        return True

    async def whose_session(self, session_id: str) -> str:
        await self._require_initialized()
        result = await execute("sessions__select_session", session_id)

        if result.empty():
            return ""
        if result.size() > 1:
            raise RuntimeError("Session ID must contain only one session")

        row = result[0]
        stored_username = row["username"].as_string()
        stored_expire_at = row["expire_at"].as_int64()

        if self._is_expired(stored_expire_at):
            await execute("sessions__remove_session", session_id)
            return ""

        return stored_username

    @staticmethod
    def _is_expired(expire_at: int) -> bool:
        return time.time() > expire_at
